File: typescript/src/TimeDepositCalculator.ts
import { TimeDeposit } from './TimeDeposit'

export class TimeDepositCalculator {
  public updateBalance(xs: TimeDeposit[]) {
    for (let i = 0; i < xs.length; i++) {
      let a = 0

      if (xs[i].days > 30) {
        if (xs[i].planType === 'student') {
          if (xs[i].days < 366) {
            a += (xs[i].balance * 0.03) / 12
          }
        } else if (xs[i].planType === 'premium') {
          if (xs[i].days > 45) {
            a += (xs[i].balance * 0.05) / 12
          }
        } else if (xs[i].planType === 'basic') {
          a += (xs[i].balance * 0.01) / 12
        }
      }

      const a2d = Math.round((a + Number.EPSILON) * 100) / 100

      xs[i].balance += a2d
    }
  };
}

File: kotlin/src/main/kotlin/org/ikigaidigital/TimeDepositCalculator.kt
package org.ikigaidigital

import java.math.BigDecimal
import java.math.RoundingMode

class TimeDepositCalculator {
    fun updateBalance(xs: List<TimeDeposit>) {
        for (i in xs.indices) {
            var interest = 0.0
            if (xs[i].days > 30) {
                if (xs[i].planType == "student") {
                    if (xs[i].days < 366) {
                        interest += xs[i].balance * 0.03 / 12
                    }
                } else if (xs[i].planType == "premium") {
                    if (xs[i].days > 45) {
                        interest += xs[i].balance * 0.05 / 12
                    }
                } else if (xs[i].planType == "basic") {
                    interest += xs[i].balance * 0.01 / 12
                }
            }
            val a2d = BigDecimal(interest).setScale(2, RoundingMode.HALF_UP)
            xs[i].balance += a2d.toDouble()
        }
    }
}